cleanup pert subroutines

weiyuan-jiang opened the pull request:
This PR addresses part 2 of issue #373 to clean up pert subroutines

gmao-rreichle commented on src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_upd_routines.F90:
```diff
@@ -3339,17 +3339,17 @@ subroutine get_obs_pert( N_ens, N_obs, N_obs_param,                  &
     
-    allocate(Obs_pert_ntrmdt(N_obs_param,pert_grid_lH%N_lon,pert_grid_lH%N_lat,N_ens))
-    allocate(Obs_pert_grid(  N_obs_param,pert_grid_lH%N_lon,pert_grid_lH%N_lat,N_ens))
+    allocate(Obs_pert_ntrmdt(pert_grid_lH%N_lon, pert_grid_lH%N_lat, N_obs_param, N_ens))
```
Here  and in the next line, do we need to allocate with "N_obs_param" when in the call to get_pert() below we're only using "N_assim_species"?
By construction, N_assim_species <= N_obs_param. 
I'm pretty sure we can save a bit of memory.  Can you please double-check?

gmao-rreichle commented:
I added a few comments and cleaned up a bit.  I am testing now.  Will update with results later.

gmao-rreichle commented on src/Components/GEOSldas_GridComp/GEOSlandpert_GridComp/land_pert.F90:
```diff
@@ -503,7 +257,7 @@ subroutine LDASsa_get_pert(                         &
           do i=1,pert_grid_l%N_lon
 
-             call adjust_mean(pert_grid_l%N_lat, N_ens, Pert_ntrmdt(m,i,:,:) )
+             call adjust_mean(pert_grid_l%N_lat, N_ens, Pert_ntrmdt(i,:,m,:) )
```
A silly question... We are passing a 4-dim array with two singleton dimensions into adjust_mean(), which expects a 2-dim array.  Does Fortran automatically "squeeze" the singleton dimensions and effectively present the desired 2-dim array to adjust_mean? 
In the new branch, the non-singleton dimension are no longer consecutive, as they are in the original code.

weiyuan-jiang replied:
It may be faster if if the two-D is the first two because of the continuous memory.  Fortran won't squeeze it but would remember the stride between elements.

gmao-rreichle commented:
@weiyuan-jiang :
I successfully tested "conus", "global" and "globalcs" with Intel.   Didn't test Gnu.
One more question: If I understand the modifications correctly, the change in the order of array dimensions shouldn't impact the pert restart files, correct?  I think the Nightly assim test uses pert restarts, so everything should be ok and tested, but I thought I should mention pert restarts just in case I misunderstood or missed something.
Besides that, perhaps we still want to change the allocation of Obs_pert_ntrmdt and Obs_pert_grid. See inline comment above.
Please review what I've changed and let me know, thanks!

weiyuan-jiang commented:
It won't affect the landpert restart order because all I changed is the 'LDASsa' style order that has nothing to do with restart. I have changed the allocation sizes of  Obs_pert_ntrmdt and Obs_pert_grid.

weiyuan-jiang commented:
The purpose to change the "LDASsa" style order is to preserve the "GEOSldas"  style order that would affect the restarts

gmao-rreichle commented:
@biljanaorescanin :
When you get a chance, please run a full test of this branch (incl GNU) and, if successful, go ahead with the merge, unless there are objections from @weiyuan-jiang.  If anything is unclear, we can discuss on Wednesday.  Thanks.

weiyuan-jiang commented:
I have tested it. So it should be good to merge. @biljanaorescanin @gmao-rreichle

biljanaorescanin commented:
All tests passed. ( it took a while discover is so slow)
